### PR TITLE
Fix setting url to http.request

### PR DESCRIPTION
**Pull request by jfilla**
Service `http.request` can be any class implementing interface IRequest. If you have custom IRequest implementation and `console.url` config is set, DI container cache looks like this: 
```php
public function createServiceHttp__request(): Flowgate\FunctionalTests\RequestMock
{
	return new Nette\Http\Request(new Nette\Http\UrlScript('http://localhost'));
}
```
, which causes type error. It should look like this
```php
public function createServiceHttp__request(): Flowgate\FunctionalTests\RequestMock
{
	return new Flowgate\FunctionalTests\RequestMock;
}
```

### CODE REVIEW

**Comment by mabar** [comment]
I believe this works in your case in tests, but it breaks behavior of ConsoleExtension for anyone who overrides Request for standard usage. Imho what should be done here is to override RequestFactory by ConsoleExtension instead so you are still able to override Request

**Comment by jfilla** [comment]
I tried what i think is standard usage. Creating custom request

**Comment by jfilla** [comment]
I tried what i think is standard usage (is it?). Creating class
```php
<?php


namespace App;


use Nette\Http\Request;
use Nette\Http\UrlScript;

class CustomRequest extends Request
{

    public function __construct()
    {
        parent::__construct(new UrlScript('http://whatever'));
    }

}
``` 
registering it as http.request
```neon
services:
    http.request: App\CustomRequest
``` 
in cli command i tried to call 
```php
$this->container->getByType(IRequest::class);
```
which fails with error
```
TypeError: Return value of Container_b2518d818a::createServiceHttp__request() must be an instance of App\CustomRequest, instance of Nette\Http\Request returned
```
because of this inside DI container
```php
public function createServiceHttp__request(): App\CustomRequest
{
	return new Nette\Http\Request(new Nette\Http\UrlScript('http://localhost'));
}
```
 I think with console.url enabled is imposible to override http.request anyway.  I will think about it and figure out better solution.. and fix tests of course.

File: src/DI/ConsoleExtension.php [hunk 1]
```diff
@@ -153,7 +153,9 @@ public function beforeCompile(): void
 		if ($config->url !== null && $builder->hasDefinition('http.request')) {
 			/** @var ServiceDefinition $httpDef */
 			$httpDef = $builder->getDefinition('http.request');
-			$httpDef->setFactory(Request::class, [new Statement(UrlScript::class, [$config->url])]);
```
Comment (f3l1x) [inline src/DI/ConsoleExtension.php:156]: What about this.

```
$httpDef->setFactory($httpDef->getFactory(), [new Statement(UrlScript::class, [$config->url])]);
```

You must preserver arguments order/position.

Comment (mabar) [reply src/DI/ConsoleExtension.php:156]: I am worried it would still break do to possibly different constructor

Comment (jfilla) [reply src/DI/ConsoleExtension.php:156]: I was thinking about that too, but i didnt like the idea of preserving argument positions. Original idea was cleaner for me.

**Comment by jfilla** [comment]
What about now? I created console requestFactory, which is used only when default requestFactory is used. Exception is thown on custom requestFactory implementation and console.url set. Console.url in combination with custom requestFactory is in my opinion useless, url should be set by custom requestFactory itself.

**Comment by mabar** [comment]
Looks way better. I will give it some final touches later today before release. Thanks

**Comment by f3l1x** [comment]
Hey @jfilla, I want to assure you we'll handle this issue. We didn't have a time yet.

**Comment by f3l1x** [comment]
Thank you @jfilla ❤️

**Pull request closed without merging**